lib/datasource/crate/index.ts

```diff
@@ -1,43 +1,168 @@
+import { join } from 'path';
+import * as fs from 'fs-extra';
+import Git from 'simple-git';
+import { logger } from '../../logger';
 import { ExternalHostError } from '../../types/errors/external-host-error';
 import * as packageCache from '../../util/cache/package';
+import { ensureCacheDir } from '../../util/fs';
 import { Http } from '../../util/http';
 import { GetReleasesConfig, Release, ReleaseResult } from '../common';
 
 export const id = 'crate';
 
+export const registryStrategy = 'first';
+
+const registryClonePaths: Record<string, string> = {};
+
 const http = new Http(id);
 
-const BASE_URL =
+const CRATES_IO_BASE_URL =
   'https://raw.githubusercontent.com/rust-lang/crates.io-index/master/';
 
-export function getIndexSuffix(lookupName: string): string {
+enum RegistryFlavor {
+  /// https://crates.io, supports rawgit access
+  CratesIo,
+
+  /// https://cloudsmith.io, needs git clone
+  Cloudsmith,
+
+  /// unknown, assuming private git repository
+  Other,
+}
+
+interface RegistryInfo {
+  flavor: RegistryFlavor;
+
+  /// raw URL of the registry, as specified in cargo config
+  rawUrl?: string;
+
+  /// parsed URL of the registry
+  url?: URL;
+
+  /// path where the registry is cloned
+  clonePath?: string;
+}
+
+export function getIndexSuffix(lookupName: string): string[] {
```

Comment:
This now returns an array of parts rather than a `/`-separated string, because it ends up being a file path, so in case this ever runs on windows, `path.join` will do the right thing.